### PR TITLE
UI - keep selection on cluster when group changes

### PR DESCRIPTION
**Pull request by guilpier-code**

### CODE REVIEW

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h [hunk 1]
```diff
@@ -76,8 +76,12 @@ class ClustersByOrder : public Yuni::IEventObserver<ClustersByOrder>, public IDa
     void onAreaChanged(Data::Area* area);
     void onInvalidateAllAreas();
 
+    void hasGroupChanged(bool b);
+    bool hasGroupChanged();
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h]: const

File: src/ui/simulator/toolbox/input/thermal-cluster.cpp [hunk 2]
```diff
@@ -174,6 +168,19 @@ void ThermalCluster::update()
     updateInnerValues();
 }
 
+void ThermalCluster::updateWhenGroupChanges()
+{
+    // Identify the selected data source (AZ or ZA ?)
+    if (pThListbox->datasource() == dsZA_)
+        dsZA_->hasGroupChanged(true);
+    else
+        dsAZ_->hasGroupChanged(true);
```
Comment (flomnes) [inline src/ui/simulator/toolbox/input/thermal-cluster.cpp]: Why not just this ?
```cpp
pThListbox->datasource()->hasGroupChanged(true);
```

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h (hunk 1 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h]: Done

File: src/ui/simulator/toolbox/input/thermal-cluster.cpp (hunk 2 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/input/thermal-cluster.cpp]: Done

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h [hunk 3]
```diff
@@ -71,13 +76,23 @@ class ClustersByOrder : public Yuni::IEventObserver<ClustersByOrder>, public IDa
...
 
     void onAreaChanged(Data::Area* area);
     void onInvalidateAllAreas();
 
+    void hasGroupChanged(bool b);
+    bool hasGroupChanged() const;
+
+private:
+    virtual void reordering_items_list(const wxString& search) = 0;
+    virtual void rebuilding_items_list(const wxString& search) = 0;
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h]: Use camelCase

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp [hunk 4]
```diff
@@ -66,78 +66,132 @@ void GetRenewableClusterMap(Data::Area* area, RenewableClusterMap& l, const wxSt
...
+                    ClusterItem* clusterItem = clusters_to_items_[*j];
+                    pParent.setElement(clusterItem, index_item);
+                    index_item++;
+                }
+            }
+        }
+    }
+}
+
+void RenewableClustersByOrder::rebuilding_items_list(const wxString& search)
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp]: camelCase

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp [hunk 5]
```diff
@@ -66,78 +66,132 @@ void GetRenewableClusterMap(Data::Area* area, RenewableClusterMap& l, const wxSt
...
+                wxString groupName = group_it->first;
+                wxString groupTitle = groupNameToGroupTitle(pArea, groupName);
+                RenewableClusterList& groupClusterList = group_it->second;
 
                 // Refreshing the group
-                pParent.add(new Antares::Component::HTMLListbox::Item::Group(s));
+                IItem* groupItem = new Group(groupTitle);
+                pParent.add(groupItem);
+                // Mapping group name to cluster item for possible further usage
+                groups_to_items_[groupName] = groupItem;
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp]: pGroupsToItems

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp [hunk 6]
```diff
@@ -66,78 +66,132 @@ void GetThermalClusterMap(Data::Area* area, ThermalClusterMap& l, const wxString
...
+        // In case the cluster group is new to the item list, we resize the list 
+        int nombreItems = sizeThermalClusterMap(l);
+        pParent.resizeTo(nombreItems);
+
         if (!l.empty())
         {
-            ThermalClusterMap::iterator end = l.end();
-            for (ThermalClusterMap::iterator group_it = l.begin(); group_it != end; ++group_it)
+            int index_item = 0;
+            for (ThermalClusterMap::iterator group_it = l.begin(); group_it != l.end(); ++group_it)
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp]: The check on list size is unnecessary : if `l` is empty, the loop will be, too.

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp [hunk 7]
```diff
@@ -66,78 +66,132 @@ void GetThermalClusterMap(Data::Area* area, ThermalClusterMap& l, const wxString
...
+                wxString groupTitle = groupNameToGroupTitle(pArea, groupName);
+                ThermalClusterList& groupClusterList = group_it->second;
 
                 // Refreshing the group
-                pParent.add(new Antares::Component::HTMLListbox::Item::Group(s));
-                
+                IItem* groupItem = new Group(groupTitle);
+                pParent.add(groupItem);
+                // Mapping group name to cluster item for possible further usage
+                groups_to_items_[groupName] = groupItem;
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp]: pGroupsToItems

File: src/ui/simulator/toolbox/input/renewable-cluster.cpp [hunk 8]
```diff
@@ -154,17 +137,15 @@ void RenewableCluster::internalBuildSubControls()
...
+    dsZA_ = pRnListbox->addDatasource<RenewableClustersByAlphaReverseOrder>();
 
     if (pAreaNotifier)
     {
-        pAreaNotifier->onAreaChanged.connect(dsAZ,
+        pAreaNotifier->onAreaChanged.connect(dsAZ_,
                                              &RenewableClustersByAlphaOrder::onAreaChanged);
         pAreaNotifier->onAreaChanged.connect(
-          dsZA, &RenewableClustersByAlphaReverseOrder::onAreaChanged);
+          dsZA_, &RenewableClustersByAlphaReverseOrder::onAreaChanged);
```
Comment (flomnes) [inline src/ui/simulator/toolbox/input/renewable-cluster.cpp]: No need for the terminal underscore

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h (hunk 3 above)
Comment (flomnes) [reply src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h]: Also, **reorder** and **rebuild**

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp [hunk 9]
```diff
@@ -66,78 +66,132 @@ void GetRenewableClusterMap(Data::Area* area, RenewableClusterMap& l, const wxSt
...
                 // Refreshing all clusters of the group
-                refreshClustersInGroup(group_it->second);
+                sortClustersInGroup(groupClusterList);
+
+                for (RenewableClusterList::iterator j = groupClusterList.begin(); j != groupClusterList.end(); ++j)
+                {
+                    ClusterItem* clusterItem = new RenewableClusterItem(*j);
+                    pParent.add(clusterItem);
+                    // Mapping real cluster to cluster item for possible further usage
+                    clusters_to_items_[*j] = clusterItem;
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp]: Remove final underscore

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h (hunk 3 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/components/htmllistbox/datasource/cluster-order.h]: ok, done now

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp (hunk 4 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/components/htmllistbox/datasource/renewable-cluster-order.cpp]: Done

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp (hunk 6 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp]: ok, done

File: src/ui/simulator/toolbox/components/htmllistbox/component.cpp [hunk 10]
```diff
@@ -147,6 +147,26 @@ void Component::add(Item::IItem::Ptr it)
...
+
+void Component::resizeTo(int newSize)
+{
+    pItems.resize(newSize);
+}
+
+void Component::force_redraw()
+{
+    pInvalidated = true;
+}
```
Comment (flomnes) [inline src/ui/simulator/toolbox/components/htmllistbox/component.cpp]: forceRedraw

File: src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp (hunk 7 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/components/htmllistbox/datasource/thermal-cluster-order.cpp]: Done

File: src/ui/simulator/toolbox/components/htmllistbox/component.cpp (hunk 10 above)
Comment (guilpier-code) [reply src/ui/simulator/toolbox/components/htmllistbox/component.cpp]: Done

**Review by flomnes — APPROVED** [review]

**Pull request merged**